### ISSUE TITLE
Safekeepers unable to connect to pageserver: `reader taken`

### ISSUE THREAD

**Issue by dvaldivia** [neondatabase/neon#1959]
I'm starting a distributed Neon DB but I'm seeing my safekeepers go into a repetitive error loop where they sort of complaint they can't connect to the pageserver.

After adding the tenant and timeline, I can see in my bucket two files being created, more over, a request every second to put the index object, so I assume connectivity to the object store is working, however when I start my compute nodes, I see the safekeepers going into the following error `ERROR connection handler exited: reader taken` then  the following

```
2022-06-19T02:55:04.709815Z ERROR {tid=17}: query handler for 'START_WAL_PUSH postgresql://(null):@pageserver-0.pageserver.default.svc.cluster.local:6400' failed: failed to run ReceiveWalConn

Caused by:
    connection closed unexpectedly
```

the page server on it's part it's throwing the following 3 log lines over and over

```
2022-06-19T02:59:17.525797Z  INFO wal_receiver_main:timeline{id=58eeeb8b6a87e7da53e899e72e21a120/95a5d23f0b82b23bcfaf198f9eb0cce3}: Switching to different safekeeper 2 for timeline 58eeeb8b6a87e7da53e899e72e21a120/95a5d23f0b82b23bcfaf198f9eb0cce3, reason: NoExistingConnection
2022-06-19T02:59:17.526102Z  INFO wal_receiver_main:timeline{id=58eeeb8b6a87e7da53e899e72e21a120/95a5d23f0b82b23bcfaf198f9eb0cce3}:safekeeper_handle{sk=2}: connecting to host=0.0.0.0 port=6500 options='-c ztimelineid=95a5d23f0b82b23bcfaf198f9eb0cce3 ztenantid=58eeeb8b6a87e7da53e899e72e21a120'
2022-06-19T02:59:17.526607Z  WARN wal_receiver_main:timeline{id=58eeeb8b6a87e7da53e899e72e21a120/95a5d23f0b82b23bcfaf198f9eb0cce3}:safekeeper_handle{sk=2}: Walreceiver connection for id 58eeeb8b6a87e7da53e899e72e21a120/95a5d23f0b82b23bcfaf198f9eb0cce3 failed with error: Failed to open walreceiver conection: error connecting to server: Connection refused (os error 111): Connection refused (os error 111)
2022-06-19T02:59:17.526743Z ERROR wal_receiver_main:timeline{id=58eeeb8b6a87e7da53e899e72e21a120/95a5d23f0b82b23bcfaf198f9eb0cce3}: WAL receiver task failed: Walreceiver connection for id 58eeeb8b6a87e7da53e899e72e21a120/95a5d23f0b82b23bcfaf198f9eb0cce3 failed with error: Failed to open walreceiver conection: error connecting to server: Connection refused (os error 111): Connection refused (os error 111), reconnecting
```

How can I debug the what is refusing the connection? 

My setup right now is:

1 pageserver
3 safekeepers
1 compute node

logs for a safekeeper

```
2022-06-19T02:54:50.953481Z  INFO version: git-env:699f46cd84c23bdfbd382679e087a5b55da87eb6
2022-06-19T02:54:50.955657Z  INFO safekeeper ID 0
2022-06-19T02:54:50.956185Z  INFO Starting safekeeper on 0.0.0.0:6500
2022-06-19T02:54:50.956294Z  INFO Auth is disabled
2022-06-19T02:54:51.021844Z  INFO WAL backup launcher started, remote config Some(RemoteStorageConfig { max_concurrent_syncs: 50, max_sync_errors: 10, storage: AwsS3(S3Config { bucket_name: "neon", bucket_region: "us-east-1", prefix_in_bucket: Some("basic"), concurrency_limit: 100 }) })
2022-06-19T02:54:51.021864Z  INFO broker: started, broker endpoints [Url { scheme: "http", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("etcd-client.default.svc.cluster.local")), port: Some(2379), path: "/", query: None, fragment: None }]
2022-06-19T02:54:51.032033Z  INFO Using s3 bucket 'neon' in region 'us-east-1' as a remote storage, prefix in bucket: 'Some("basic")', bucket endpoint: 'Some("ns-4.cloud.min.dev")'
2022-06-19T02:54:51.068809Z  INFO broker: Subscribing to broker value updates, key: SubscriptionKey { cluster_prefix: "neon", kind: All }
2022-06-19T02:54:51.102973Z  INFO Starting an HTTP endpoint at 0.0.0.0:7676
2022-06-19T02:55:04.594263Z  INFO {tid=17}: got query "START_WAL_PUSH postgresql://(null):@pageserver-0.pageserver.default.svc.cluster.local:6400"
2022-06-19T02:55:04.594909Z  INFO {tid=17}: loading control file ./58eeeb8b6a87e7da53e899e72e21a120/95a5d23f0b82b23bcfaf198f9eb0cce3/safekeeper.control
2022-06-19T02:55:04.600888Z  INFO {tid=17}: timeline 95a5d23f0b82b23bcfaf198f9eb0cce3 restored
2022-06-19T02:55:04.602552Z  INFO {tid=17}: find_end_of_wal_segment reached zeros at 0/1696158, last records ends at 0/1696158    
2022-06-19T02:55:04.602763Z  INFO {tid=17}: initialized storage for timeline 95a5d23f0b82b23bcfaf198f9eb0cce3, flush_lsn=0/1696158, commit_lsn=0/1696158, peer_horizon_lsn=0/16960E8
2022-06-19T02:55:04.605480Z  INFO {tid=17}:WAL acceptor{timeline=95a5d23f0b82b23bcfaf198f9eb0cce3}: start handshake with wal proposer 10.244.2.35:48796 sysid 0 timeline 1
2022-06-19T02:55:04.609366Z  INFO {tid=17}:WAL acceptor{timeline=95a5d23f0b82b23bcfaf198f9eb0cce3}: processed greeting from proposer [137, 36, 155, 182, 62, 174, 177, 45, 88, 213, 183, 131, 29, 11, 231, 161], sending term 11
2022-06-19T02:55:04.609963Z  INFO {tid=17}:WAL acceptor{timeline=95a5d23f0b82b23bcfaf198f9eb0cce3}: timeline 58eeeb8b6a87e7da53e899e72e21a120/95a5d23f0b82b23bcfaf198f9eb0cce3 active=true now
2022-06-19T02:55:04.611148Z  INFO starting WAL backup task for 58eeeb8b6a87e7da53e899e72e21a120/95a5d23f0b82b23bcfaf198f9eb0cce3
2022-06-19T02:55:04.613102Z  INFO WAL backup task{zttid=58eeeb8b6a87e7da53e899e72e21a120/95a5d23f0b82b23bcfaf198f9eb0cce3}: started
2022-06-19T02:55:04.614380Z  INFO WAL backup task{zttid=58eeeb8b6a87e7da53e899e72e21a120/95a5d23f0b82b23bcfaf198f9eb0cce3}: acquiring leadership
2022-06-19T02:55:04.619111Z  INFO {tid=17}:WAL acceptor{timeline=95a5d23f0b82b23bcfaf198f9eb0cce3}: processed VoteRequest for term 12: VoteResponse { term: 12, vote_given: 1, flush_lsn: 0/1696158, truncate_lsn: 0/16960E8, term_history: [(11, 0/1696158), (10, 0/1696158), (9, 0/1696158), (8, 0/1696158), (7, 0/1696158), (6, 0/1696158), (5, 0/1696158), (4, 0/1696158), (3, 0/1696158), (2, 0/16960E8)], timeline_start_lsn: 0/16960E8 }
2022-06-19T02:55:04.633920Z  INFO {tid=20}: got query "START_REPLICATION 0/16960E8 TIMELINE 1"
2022-06-19T02:55:04.637295Z  INFO {tid=20}:WAL sender{timeline=95a5d23f0b82b23bcfaf198f9eb0cce3}: Start replication from 0/16960E8 till Some(0/1696158)
2022-06-19T02:55:04.650415Z ERROR connection handler exited: reader taken
2022-06-19T02:55:04.654839Z  INFO {tid=17}:WAL acceptor{timeline=95a5d23f0b82b23bcfaf198f9eb0cce3}: received ProposerElected ProposerElected { term: 12, start_streaming_at: 0/1696158, term_history: [(12, 0/1696158), (11, 0/1696158), (10, 0/1696158), (9, 0/1696158), (8, 0/1696158), (7, 0/1696158), (6, 0/1696158), (5, 0/1696158), (4, 0/1696158), (3, 0/1696158), (2, 0/16960E8)], timeline_start_lsn: 0/16960E8 }
2022-06-19T02:55:04.703886Z  INFO {tid=17}:WAL acceptor{timeline=95a5d23f0b82b23bcfaf198f9eb0cce3}: start receiving WAL since 0/1696158
2022-06-19T02:55:04.708524Z  INFO stopping WAL backup task for 58eeeb8b6a87e7da53e899e72e21a120/95a5d23f0b82b23bcfaf198f9eb0cce3
2022-06-19T02:55:04.709815Z ERROR {tid=17}: query handler for 'START_WAL_PUSH postgresql://(null):@pageserver-0.pageserver.default.svc.cluster.local:6400' failed: failed to run ReceiveWalConn

Caused by:
    connection closed unexpectedly
2022-06-19T02:55:04.713656Z  INFO WAL backup task{zttid=58eeeb8b6a87e7da53e899e72e21a120/95a5d23f0b82b23bcfaf198f9eb0cce3}: task canceled

```

**Comment by dvaldivia** [comment]
perhaps the line that reports a query with a `(null)` user is telling something?

```
INFO {tid=17}: got query "START_WAL_PUSH postgresql://(null):@pageserver-0.pageserver.default.svc.cluster.local:6400"
```

**Comment by petuhovskiy** [comment]
Hi! It's generally hard to say what exactly has gone wrong, but I'll try to guess the probable issue and explain logs a little bit.

The issue is in the safekeeper configuration. More specifically, you need to set `--listen-pg` in cmd args to the safekeeper's `address:port`, which should be available for the connection from the pageserver. Currently, it's set to `0.0.0.0:6500` and pageserver tries to connect to that address, but gets `Connection refused (os error 111)`. To fix the issue, launch safekeepers with something like `--listen-pg sk-1.local:6500`, but using real address of the safekeeper instead of `sk-1.local`.

Comments on the logs:
> ```
> 2022-06-19T02:55:04.709815Z ERROR {tid=17}: query handler for 'START_WAL_PUSH postgresql://(null):@pageserver-0.pageserver.default.svc.cluster.local:6400' failed: failed to run ReceiveWalConn
> 
> Caused by:
>     connection closed unexpectedly
> ```
This message means that compute->safekeeper connection has closed. It's misleading and we're going to fix this in #635

> ```
> 2022-06-19T02:59:17.525797Z  INFO wal_receiver_main:timeline{id=58eeeb8b6a87e7da53e899e72e21a120/95a5d23f0b82b23bcfaf198f9eb0cce3}: Switching to different safekeeper 2 for timeline 58eeeb8b6a87e7da53e899e72e21a120/95a5d23f0b82b23bcfaf198f9eb0cce3, reason: NoExistingConnection
> 2022-06-19T02:59:17.526102Z  INFO wal_receiver_main:timeline{id=58eeeb8b6a87e7da53e899e72e21a120/95a5d23f0b82b23bcfaf198f9eb0cce3}:safekeeper_handle{sk=2}: connecting to host=0.0.0.0 port=6500 options='-c ztimelineid=95a5d23f0b82b23bcfaf198f9eb0cce3 ztenantid=58eeeb8b6a87e7da53e899e72e21a120'
> 2022-06-19T02:59:17.526607Z  WARN wal_receiver_main:timeline{id=58eeeb8b6a87e7da53e899e72e21a120/95a5d23f0b82b23bcfaf198f9eb0cce3}:safekeeper_handle{sk=2}: Walreceiver connection for id 58eeeb8b6a87e7da53e899e72e21a120/95a5d23f0b82b23bcfaf198f9eb0cce3 failed with error: Failed to open walreceiver conection: error connecting to server: Connection refused (os error 111): Connection refused (os error 111)
> 2022-06-19T02:59:17.526743Z ERROR wal_receiver_main:timeline{id=58eeeb8b6a87e7da53e899e72e21a120/95a5d23f0b82b23bcfaf198f9eb0cce3}: WAL receiver task failed: Walreceiver connection for id 58eeeb8b6a87e7da53e899e72e21a120/95a5d23f0b82b23bcfaf198f9eb0cce3 failed with error: Failed to open walreceiver conection: error connecting to server: Connection refused (os error 111): Connection refused (os error 111), reconnecting
> ```
We have no connections from safekeepers to pageserver, but instead pageserver connects to the best safekeeper. This error message tells that pageserver cannot connect to`0.0.0.0`, which suggests the issue.


> logs for a safekeeper

Logs look generally ok, as we can see that compute started successfully. `ERROR connection handler exited: reader taken` is also a misleading error because it just means that replication stream was closed (we should also fix this in #635). But it's strange that compute->safekeeper connection finished so early, usually it lives much longer. I suggest you check your compute configuration, it should have correct pageserver address.

**Comment by arssher** [comment]
`ERROR connection handler exited: reader taken` is fixed, otherwise issue is stale, closing.
